app/services/sso/openid_connect.rb

```diff
@@ -20,6 +22,16 @@ def authenticated?
       self.user = User.current.presence || authenticate!
     end
 
+    # Users can login using OpenId Connect in two way, one by using OpenID provider directly,
+    # second, by using the mod_auth_openidc. While using mod_auth_openidc user needs to clear
+    # the session form OpenID provider as well as mod_auth_openidc session stored in the cookie.
+    def logout_url
+      default_logout_url = Setting['login_delegation_logout_url']
+      return sso_redirect_url(CGI.escape(default_logout_url)) unless mod_auth_openidc_session_exist?
+      logout_path = CGI.escape(default_logout_url + "/redirect_uri?logout=" + CGI.escape(default_logout_url))
```

Comment:
this looks like hardcoding part of the path and would be confusing to users who would likely input the full url for logging out in the setting.

Reply:
also, why is the url added twice and escaped twice in the second time?

Reply:
> this looks like hardcoding part of the path and would be confusing to users who would likely input the full url for logging out in the setting.

@tbrisker hardcoding is done because it is according to the protocol and will not be changed. https://github.com/zmartzone/mod_auth_openidc/wiki/OpenID-Connect-Session-Management#logout

> also, why is the url added twice and escaped twice in the second time?

We need to logout from both OpenId provider (Keycloak in my usecase), and from the mod_auth_openidc. That is the reason we need to escape twice.

Reply:
> @tbrisker hardcoding is done because it is according to the protocol and will not be changed. https://github.com/zmartzone/mod_auth_openidc/wiki/OpenID-Connect-Session-Management#logout

That particular line states:
> An example logout link:

I don't see anything stating it MUST (in the RFC 2119 MUST sense) be `redirect_uri`.

To me it reads that `mod_auth_openidc` will actually do the HTTP request to keycloak to perform the logout.

Reply:
This is correct, with this in mind, now we do not need to escape twice :) thanks @ekohl